src/modules/Footer.jsx

```diff
@@ -0,0 +1,269 @@
+    while (newLine.length < line.length) {
+      newLine.push(0);
+    }
+    return direction ? newLine.reverse() : newLine;
+  };
+
+  const getRow = (array, i) => array.filter((elem, id) => id >= i && id < i + boardSize);
+
+  const getColumn = (array, i) => array.filter((elem, id) => (id - i) % boardSize === 0);
+
+  const updateNewBoardHorizontal = (newBoard, i, direction = false) => {
+    const newLine = processLine(getRow(board, i), direction);
+    newLine.map((elem, id) => {
+      // eslint-disable-next-line no-param-reassign
+      newBoard[id + i] = elem;
+      return elem;
+    });
+  };
+
+  const updateNewBoardVertical = (newBoard, i, direction = false) => {
+    const newLine = processLine(getColumn(board, i), direction);
+    newLine.map((elem, id) => {
+      // eslint-disable-next-line no-param-reassign
+      newBoard[id * boardSize + i] = elem;
+      return elem;
+    });
+  };
+
+  const moveLeft = () => {
+    const newBoard = new Array(squareBoardSize);
+    for (let i = 0; i < squareBoardSize; i += boardSize) {
+      updateNewBoardHorizontal(newBoard, i);
+    }
+    setNewBoard(newBoard);
+  };
+
+  const moveRight = () => {
+    const newBoard = new Array(squareBoardSize);
+    for (let i = 0; i < squareBoardSize; i += boardSize) {
+      updateNewBoardHorizontal(newBoard, i, true);
+    }
+    setNewBoard(newBoard);
+  };
+
+  const moveUp = () => {
+    const newBoard = new Array(squareBoardSize);
+    for (let i = 0; i < boardSize; i++) {
+      updateNewBoardVertical(newBoard, i);
+    }
+    setNewBoard(newBoard);
+  };
+
+  const moveDown = () => {
+    const newBoard = new Array(squareBoardSize);
+    for (let i = 0; i < boardSize; i++) {
+      updateNewBoardVertical(newBoard, i, true);
+    }
+    setNewBoard(newBoard);
+  };
+
+  const isGameOver = (array) => {
+    const emptySpots = array.filter((elem) => !elem).length;
+    if (emptySpots) return false;
+    for (let i = 0; i < squareBoardSize; i += boardSize) {
+      const line = getRow(array, i);
+      for (let j = 0; j < line.length - 1; j++) {
+        if (line[j] === line[j + 1]) {
+          return false;
+        }
+      }
+    }
+
+    for (let i = 0; i < boardSize; i++) {
+      const line = getColumn(array, i);
+      for (let j = 0; j < line.length - 1; j++) {
+        if (line[j] === line[j + 1]) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  };
+
+  const handleEvent = (e) => {
+    if (e.stopPropagation) e.stopPropagation();
+    if (e.code === 'ArrowLeft' || e.code === 'KeyA') moveLeft();
+    if (e.code === 'ArrowRight' || e.code === 'KeyD') moveRight();
+    if (e.code === 'ArrowUp' || e.code === 'KeyW') moveUp();
+    if (e.code === 'ArrowDown' || e.code === 'KeyS') moveDown();
+  };
+
+  useEffect(() => {
+    document.addEventListener('keydown', handleEvent);
+    return () => {
+      document.removeEventListener('keydown', handleEvent);
+    };
+  }, ['keydown', handleEvent]);
```

Comment:
why 'keydown' is here? it's just a string which doesn't make sense in the dependency array. you can set it as an empty array

Reply:
yeah, I can rid of 'keydown', though I can't remove handleEvent from dependencies, it won't work properly without it, not 100% sure why. I think it maybe connected to rerender trigered by one of move functions in handleEvent.

Reply:
In the dependency array you should add variables and functions (though it might be difficult sometimes) you used inside useEffect, or use it empty if you want to set it on didMount only and remove subscriptions from another variables. AFAIR, handleEvent is a variable :) you can change it to be a ‘function’ instead.